src/resolution/composite-arns-resolver.test.ts

```diff
@@ -0,0 +1,265 @@
+      maxConcurrentResolutions: 2,
+      resolverTimeoutMs: 50,
+    });
+
+    const result = await compositeResolver.resolve({ name: 'test.ar' });
+
+    assert.strictEqual((resolver1.resolve as any).mock.calls.length, 1);
+    assert.strictEqual((resolver2.resolve as any).mock.calls.length, 1);
+    assert.deepEqual(result, mockResolution);
+  });
+
+  it('should return undefined resolution when all resolvers fail', async () => {
+    const resolver1: NameResolver = {
+      resolve: mock.fn(async () => {
+        throw new Error('Failed 1');
+      }),
+    };
+
+    const resolver2: NameResolver = {
+      resolve: mock.fn(async () => {
+        throw new Error('Failed 2');
+      }),
+    };
+
+    const compositeResolver = new CompositeArNSResolver({
+      log,
+      resolvers: [resolver1, resolver2],
+      resolutionCache: mockResolutionCache,
+      registryCache: mockRegistryCache,
+      arnsNamesCache: {} as ArNSNamesCache, // unused, avoids unawaited promises
+      maxConcurrentResolutions: 2,
+      resolverTimeoutMs: 50,
+    });
+
+    const result = await compositeResolver.resolve({ name: 'test.ar' });
+
+    assert.deepEqual(result, {
+      name: 'test.ar',
+      resolvedId: undefined,
+      resolvedAt: undefined,
+      ttl: undefined,
+      processId: undefined,
+      limit: undefined,
+      index: undefined,
+    });
+  });
+
+  it('should respect maxConcurrentResolutions limit', async () => {
+    let activeResolutions = 0;
+    let maxActiveResolutions = 0;
+    let resolversCalled = 0;
+    const totalResolvers = 4;
+
+    const createResolver = (): NameResolver => ({
+      resolve: mock.fn(async () => {
+        resolversCalled++;
+        activeResolutions++;
+        maxActiveResolutions = Math.max(
+          maxActiveResolutions,
+          activeResolutions,
+        );
+
+        await Promise.resolve();
+
+        activeResolutions--;
+        return {
+          name: 'test.ar',
+          resolvedId: undefined,
+          resolvedAt: undefined,
+          ttl: undefined,
+          processId: undefined,
+          limit: undefined,
+          index: undefined,
+        };
+      }),
+    });
+
+    const resolvers = Array(totalResolvers)
+      .fill(null)
+      .map(() => createResolver());
+
+    const compositeResolver = new CompositeArNSResolver({
+      log,
+      resolvers,
+      resolutionCache: mockResolutionCache,
+      registryCache: mockRegistryCache,
+      arnsNamesCache: {} as ArNSNamesCache, // unused, avoids unawaited promises
+      maxConcurrentResolutions: 2,
+      resolverTimeoutMs: 50,
+    });
+
+    await compositeResolver.resolve({ name: 'test.ar' });
+
+    assert.equal(
+      maxActiveResolutions,
+      2,
+      'Max concurrent resolutions exceeded limit',
+    );
+    assert.equal(
+      resolversCalled,
+      totalResolvers,
+      'Not all resolvers were called',
+    );
+  });
+});
```

Comment:
nit: add a test that supports the `AbortSignal` aborting the request before resolution occurs